src/universal/redux/authDuck.js

```diff
@@ -31,48 +32,59 @@ export default function reducer(state = initialState, action = {}) {
   }
 }
 
-export function setAuthToken(authToken, metaProfile = {}) {
-  const profile = metaProfile || {};
-  if (!authToken) {
-    throw new Error('setAuthToken action created with undefined authToken');
-  }
-  let obj = null;
-  try {
-    obj = jwtDecode(authToken);
-  } catch (e) {
-    throw new Error(`unable to decode jwt: ${e}`);
-  }
-  cashay.create({httpTransport: new ActionHTTPTransport(authToken)});
-  if (typeof __PRODUCTION__ !== 'undefined' && __PRODUCTION__) {
-    /*
-     * Sentry error reporting meta-information. Raven object is set via SSR.
-     * See server/Html.js for how this is initialized
-     */
-    Raven.setUserContext({ // eslint-disable-line no-undef
-      id: obj.sub,
-      email: profile.email
-    });
-  }
-  return {
-    type: SET_AUTH_TOKEN,
-    payload: {
-      obj,
-      token: authToken
-    },
-    meta: {
-      analytics: {
-        eventType: EventTypes.identify,
-        eventPayload: {
-          userId: obj.sub,
-          traits: {
-            avatar: profile.picture,
-            createdAt: profile.created_at,
-            email: profile.email,
-            name: profile.name
+export function setAuthToken(authToken, newProfile) {
+  return (dispatch, getState) => {
+    const {profile: cachedProfile} = getState();
```

Comment:
While doing this implementation I discovered a problem where the socket was initiating a later call to setAuthToken and smoking the profile information coming from auth0. I could have handled this in two ways: (1) if the profile === `undefined`, opting to dispatch the `SET_AUTH_TOKEN` event without the segment meta-information or (2) caching it somewhere and always dispatching the same kind of event.

I ended up ruling out (1) when I discovered that we are only ever provided the user profile information when the user logs in. If a user closes their browser, then reopens it later and navigates to Action we'd have no way of automatically including the profile information with the segment events since we allow them to access the app as long as they hold a valid token.

A nice side-effect of storing the profile in a duck is that it is serialized to storage. When the token is restored and `setAuthToken` dispatched, segment automatically gets refreshed with the correct identity information and meta information to provide a rich event stream.